scipy/stats/stats.py

```diff
@@ -3029,28 +3057,51 @@ def pearsonr(x, y):
     >>> stats.pearsonr([1,2,3,4,5], [5,6,7,8,7])
     (0.83205029433784372, 0.080509573298498519)
     """
-    # x and y should have same length.
+    n = len(x)
+    if n != len(y):
+        raise ValueError('x and y must have the same length.')
+
+    if n < 2:
+        raise ValueError('x and y must have length at least 2.')
+
     x = np.asarray(x)
     y = np.asarray(y)
-    n = len(x)
-    mx = x.mean()
-    my = y.mean()
-    xm, ym = x - mx, y - my
-    r_num = np.add.reduce(xm * ym)
-    r_den = np.sqrt(_sum_of_squares(xm) * _sum_of_squares(ym))
-    r = r_num / r_den
+
+    # If an input is constant, the correlation coefficient is not defined.
+    if (x == x[0]).all() or (y == y[0]).all():
+        warnings.warn(PearsonRConstantInputWarning())
+        return np.nan, np.nan
+
+    if n == 2:
+        return np.float64(np.sign(x[1] - x[0])*np.sign(y[1] - y[0])), 1.0
+
+    xmean = x.mean(dtype=np.float64)
+    ymean = y.mean(dtype=np.float64)
+
+    xm = x - xmean
+    ym = y - ymean
+
+    # Unlike np.linalg.norm or the expression sqrt((xm*xm).sum()),
+    # scipy.linalg.norm(xm) does not overflow if xm is, for example,
+    # [-5e210, 5e210, 3e200, -3e200]
+    normxm = linalg.norm(xm)
+    normym = linalg.norm(ym)
+
+    threshold = 1e-13
+    if normxm < threshold*abs(xmean) or normym < threshold*abs(ymean):
+        # If all the values in x (likewise y) are very close to the mean,
+        # the loss of precision that occurs in the subtraction xm = x - xmean
+        # might result in large errors in r.
+        warnings.warn(PearsonRNearConstantInputWarning())
+
+    r = np.dot(xm/normxm, ym/normym)
 
     # Presumably, if abs(r) > 1, then it is only some small artifact of
     # floating point arithmetic.
     r = max(min(r, 1.0), -1.0)
-    df = n - 2
-    if abs(r) == 1.0:
-        prob = 0.0
-    else:
-        t_squared = r**2 * (df / ((1.0 - r) * (1.0 + r)))
-        prob = special.betainc(
-            0.5*df, 0.5, np.fmin(np.asarray(df / (df + t_squared)), 1.0)
-        )
+
+    halfdf = n/2 - 1
+    prob = 2*special.btdtr(halfdf, halfdf, 0.5*(1 - abs(r)))
```

Comment:
how does this relate to the t-distribution before?
https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#Testing_using_Student's_t-distribution

Reply:
For the null hypothesis that the population correlation ρ is 0, the distribution of r is the beta distribution on [-1, 1] with equal shape parameters a = b = n/2 - 1.  (This is the result that "Student" reported in 1908; there is no mention of the t distribution in that paper.)  The distribution is referred to as the exact distribution in the wikipedia article that you linked to (look in the subsection titled "Using the exact distribution", immediately after the subsection that you linked to).  The code in this pull request uses the exact distribution.

The t-distribution comes from the fact that t = r*sqrt((n-2)/(1-r**2)) is t-distributed under the same null hypothesis.  There is a connection between the t-distribution and the beta distribution, and the old code used that connection to compute the p-value.  The new code is mathematically equivalent to the old code.

(Note that the old code is needlessly complicated: it computes t_squared, and then passes df/(df+t_squared) to `betainc`.  But if you plug the expression for t_squared into df/(df+t_squared), it simplifies to 1-r**2!)

I didn't touch the docstring in this pull request, but I will.  I realize this might make https://github.com/scipy/scipy/pull/9498 obsolete.  (@rlucas7, sorry I keep stepping on your pull requests!)

In the docstring, I can say that the p-value is computed with the exact distribution of r, and give a brief explanation of what that distribution is.  I'll also add comments in the code to explain exactly where `2*special.btdtr(halfdf, halfdf, 0.5*(1 - abs(r)))` comes from.

Reply:
>I didn't touch the docstring in this pull request, but I will. I realize this might make #9498 >obsolete. (@rlucas7, sorry I keep stepping on your pull requests!)

@WarrenWeckesser no worries, I'm glad the method is fixed, and I prefer your solution which addresses all of the issues in 1 PR. I'm all about better code, not getting more PRs in :).

>n the docstring, I can say that the p-value is computed with the exact distribution of r, and >give a brief explanation of what that distribution is.

Here I would recommend to use something other than the word "exact" because "exact" usually means a small sample permutation test where the code generates all possible permutations and calculates a test statistic (or p-value) from the-empirically generated-distribution.

Also, this is only the correct distribution if the observations come from normal distributions. The observations may not come from normals and then this would be an asymptotic approximation (assuming the distribution that generates the data has finite mean and variance so CLT can kick in).

Reply:
@rlucas7, take a look at the new Notes section of the docstring, where, after giving the formula for the PDF f(r), I say "This is sometimes referred to as the exact distribution of r."  The distribution is "exact" in the sense that it is not an asymptotic result.  I know that "exact" is also used in other contexts (e.g. Fisher's exact test), and the overloading of the meaning is unfortunate.  I'm sure I have seen this use of "exact" in reference to this distribution for Pearson's r more than once, but at the moment the only reference that I can find is the wikipedia page: https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#Using_the_exact_distribution

Regarding the normal assumption: also in the Notes section, before giving the formula for the PDF, I say "Under the assumption that x and y are drawn from independent normal distributions (so the population correlation coefficient is 0) ..."

Reply:
Sounds like @rlucas7 is +1 on the PR below & the docs have the appropriate content now, so marking this as resolved